File: src/Lessons/2-Arrays/OddOccurencesInArray.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	//Return 10
	fmt.Println(solution([]int{2, 2, 4, 4, 5, 6, 6, 5, 7, 8, 9, 10, 7, 8, 9}))

}

func solution(someArray []int) int {
	//get the unpaired element
	if len(someArray) == 1 {
		//if only one item in array
		return someArray[0]
	}
	//sort our array
	sort.Ints(someArray)
	//loop through
	oddNumber := 0
	//since numbers are in pairs we loop increasing our index by 2
	for index := 1; index < len(someArray); index += 2 {
		number := someArray[index]
		previousNumber := someArray[index-1]
		if number != previousNumber {
			oddNumber = previousNumber
			break
		}
	}
	// if no oddnumber the oddnumber is then the last number in the array
	if oddNumber == 0 {
		oddNumber = someArray[len(someArray)-1]
	}
	return oddNumber
}

File: src/Lessons/2-Arrays/OddOccurencesInArray.js
//Odd Occurences in Array
function solution(A) {
    //Return integer within Array with no pair
    let integerArray = A;
    //if array contains only one number return the number
    if(integerArray.length == 1) return A[0];
    //First sort our array
    integerArray.sort();
    //Loop through
    let oddNumber = 0;
    for (let index = 1; index < integerArray.length; index+=2) {
        const number = integerArray[index];
        const prevNumber = integerArray[index-1];

        if(number != prevNumber){
            oddNumber = prevNumber;
            break;
        }
    }
    //incase the last number in the array is the odd one
    if(oddNumber == 0) oddNumber = integerArray[integerArray.length-1];
    return oddNumber;
}

console.log(solution([2, 2, 3, 3, 4,4,6,6,7]))